src/Export/ProductService.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+declare(strict_types=1);
+
+namespace FINDOLOGIC\FinSearch\Export;
+
+use FINDOLOGIC\FinSearch\Utils\Utils;
+use Psr\Container\ContainerInterface;
+use Shopware\Core\Content\Product\Aggregate\ProductVisibility\ProductVisibilityDefinition;
+use Shopware\Core\Content\Product\SalesChannel\ProductAvailableFilter;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\EntitySearchResult;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\MultiFilter;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\IdSearchResult;
+use Shopware\Core\Framework\Uuid\Uuid;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+
+class ProductService
+{
+    public const CONTAINER_ID = 'fin_search.product_service';
+
+    /** @var ContainerInterface */
+    private $container;
+
+    /** @var SalesChannelContext|null */
+    private $salesChannelContext;
+
+    public static function getInstance(
+        ContainerInterface $container,
+        ?SalesChannelContext $salesChannelContext
+    ): ProductService {
+        if ($container->has(self::CONTAINER_ID)) {
+            $productService = $container->get(self::CONTAINER_ID);
+        } else {
+            $productService = new ProductService($container, $salesChannelContext);
+            $container->set(self::CONTAINER_ID, $productService);
+        }
+
+        if ($salesChannelContext && !$productService->getSalesChannelContext()) {
+            $productService->setSalesChannelContext($salesChannelContext);
+        }
+
+        return $productService;
+    }
+
+    public function __construct(ContainerInterface $container, ?SalesChannelContext $salesChannelContext = null)
+    {
+        $this->container = $container;
+        $this->salesChannelContext = $salesChannelContext;
+    }
+
+    public function setSalesChannelContext(SalesChannelContext $salesChannelContext): void
+    {
+        $this->salesChannelContext = $salesChannelContext;
+    }
+
+    public function getSalesChannelContext(): ?SalesChannelContext
+    {
+        return $this->salesChannelContext;
+    }
+
+    public function getTotalProductCount(): int
+    {
+        $criteria = $this->getCriteriaWithProductVisibility();
+
+        /** @var IdSearchResult $result */
+        $result = $this->container->get('product.repository')->searchIds(
+            $criteria,
+            $this->salesChannelContext->getContext()
+        );
+
+        return $result->getTotal();
+    }
+
+    public function searchVisibleProducts(
+        ?int $limit = null,
+        ?int $offset = null,
```

Comment:
This service is clear at a database level, so I explicitly chose to use `$limit` and `$offset`, because it could be used somewhere else in the application as well. It should not be tight with the exporter.